src/main/java/org/truffleruby/core/array/ArraySliceNode.java

```diff
@@ -42,9 +42,9 @@ public DynamicObject readInBounds(DynamicObject array,
             return createArray(null, 0);
         } else {
             final ArrayMirror arrayMirror = strategy.newMirror(array);
-            final Object store = arrayMirror.extractRange(from, to).getArray();
-            strategy.setStore(array, arrayMirror.extractRange(0, strategy.getSize(array)).getArray());
-            return createArray(store, to - from);
+            final Object slice = arrayMirror.extractRange(from, to).getArray();
+            strategy.makeStorageImmutable(array);
```

Comment:
Is this the same as "unsharing" the array storage?

Reply:
I think this kind of pattern might be clearer if it works too:
```
final Object slice = arrayMirror.extractRange(from, to).copyArrayAndMirror().getArray();
# Or
final Object slice = arrayMirror.extractRange(from, to).unshareStorage().getArray();
# Or
final Object slice = arrayMirror.extractRange(from, to).getUnsharedArray();
```
Immutable seems to carry the wrong meaning for me, that storage can still be mutated of course.